common/src/main/java/com/microsoft/identity/common/internal/cache/HelloCache.java

```diff
@@ -0,0 +1,152 @@
+        mProtocolName = protocolName;
+        mTargetAppPackageName = targetAppPackageName;
+    }
+
+    /**
+     * Gets the cached negotiated protocol version. Returns null if there is none.
+     *
+     * @param clientMinimumProtocolVersion minimum version of the protocol that the client supports.
+     * @param clientMaximumProtocolVersion maximum version of the protocol that to be advertised by the client.
+     */
+    public @Nullable String tryGetNegotiatedProtocolVersion(@Nullable final String clientMinimumProtocolVersion,
+                                                            @NonNull final String clientMaximumProtocolVersion) {
+        final String methodName = ":tryGetNegotiatedProtocolVersion";
+
+        final String key;
+        try {
+            key = getKey(clientMinimumProtocolVersion, clientMaximumProtocolVersion);
+        } catch (final PackageManager.NameNotFoundException e) {
+            Logger.error(TAG + methodName, "Failed to retrieve key", e);
+            return null;
+        }
+
+        return mFileManager.getString(key);
+    }
+
+    /**
+     * Store the given negotiated protocol version into the cache.
+     *
+     * @param clientMinimumProtocolVersion minimum version of the protocol that the client supports.
+     * @param clientMaximumProtocolVersion maximum version of the protocol that to be advertised by the client.
+     * @param negotiatedProtocolVersion    the negotiated protocol version as returned from hello().
+     */
+    public void saveNegotiatedProtocolVersion(@Nullable final String clientMinimumProtocolVersion,
+                                              @NonNull final String clientMaximumProtocolVersion,
+                                              @NonNull final String negotiatedProtocolVersion) {
+        final String methodName = ":saveNegotiatedProtocolVersion";
+
+        final String key;
+        try {
+            key = getKey(clientMinimumProtocolVersion, clientMaximumProtocolVersion);
+        } catch (final PackageManager.NameNotFoundException e) {
+            Logger.error(TAG + methodName, "Failed to retrieve key", e);
+            return;
+        }
+
+        mFileManager.putString(key, negotiatedProtocolVersion);
+    }
+
+    /**
+     * Generates {@link SharedPreferencesFileManager}'s key.
+     *
+     * @param clientMinimumProtocolVersion minimum version of the protocol that the client supports.
+     * @param clientMaximumProtocolVersion maximum version of the protocol that to be advertised by the client.
+     */
+    private @NonNull String getKey(@Nullable final String clientMinimumProtocolVersion,
+                                   @NonNull final String clientMaximumProtocolVersion)
+            throws PackageManager.NameNotFoundException {
+        return mProtocolName +
+                "[" + clientMinimumProtocolVersion + "," + clientMaximumProtocolVersion + "]:"
+                + mTargetAppPackageName + "[" + getVersionCode() + "]";
```

Comment:
this cache is actually in the calling app itself right? If yes, then do we even need to store the package name?

Reply:
package name of the broker.

Reply:
Got it